network-libp2p: Filter valid addresses and fix discovery protocol

jsdanielh opened the pull request:
- Add several changes to filter valid addresses:
  - Remove the `identify` behaviour since it was returning peer addresses that sometimes are not valid such as secure websocket and IPs multiaddresses.
  - Let know the discovery behaviour about new listen addresses to mimic the `identify` behaviour and be able to discover more of our own addresses.
  - Add some filtering in the `PeerContactBook` such that the `discovery` and the `connection_pool` behaviours only return valid addresses (in `handle_pending_outbound_connection`).
  - Add a configuration knob in the network to allow only secure websocket connections and another one for allowing loopback addresses.
  - Only add an address for a peer in the DHT behaviour if it passes the validity filters such that it doesn't return invalid addresses when the swarm dials a peer.
  This fixes #2052.
- Fix the discovery protocol to always periodically send our own `PeerContact` (along with some of our contacts) which is updated frequently to avoid being aged.

## Pull request checklist

- [X] All tests pass. The project builds and runs.
- [X] I have resolved any merge conflicts.
- [X] I have resolved all `clippy` and `rustfmt` warnings.

jsdanielh committed:
network-libp2p: Filter valid addresses

Add several changes to filter valid addresses:
- Remove the `identify` behaviour since it was returning peer
  addresses that sometimes are not valid such as secure websocket and
  IPs multiaddresses.
- Let know the discovery behaviour about new listen addresses to
  mimic the `identify` behaviour and be able to discover more of our
  own addresses.
- Add some filtering in the `PeerContactBook` such that the
  `discovery` and the `connection_pool` behaviours only return valid
  addresses (in `handle_pending_outbound_connection`).
- Add a configuration knob in the network to allow only secure
  websocket connections and another one for allowing loopback
  addresses.
- Only add an address for a peer in the DHT behaviour if it passes
  the validity filters such that it doesn't return invalid addresses
  when the swarm dials a peer.

This fixes #2052.

jsdanielh committed:
network-libp2p: Fix discovery protocol

Fix the discovery protocol to always periodically send our own
`PeerContact` (along with some of our contacts) which is updated
frequently to avoid being aged.

hrxi commented on network-libp2p/src/discovery/handler.rs:
```diff
@@ -210,14 +210,18 @@ impl Handler {
+    /// This list also includes our own contact which should be already filtered since we already have
+    /// connections to these peers.
+    fn get_peer_contacts(
+        &self,
```
```suggestion
```
No longer necessary.

jsdanielh replied:
It's still required for `self.service_filter`

hrxi replied:
Oops.

hrxi commented on network-libp2p/src/config.rs:
```diff
@@ -58,7 +62,7 @@ impl Config {
                 message.topic.hash(&mut s);
                 message.data.hash(&mut s);
-                gossipsub::MessageId::from(s.finish().to_string())
+                gossipsub::MessageId::from(s.finish().to_be_bytes())
```
Unrelated to this PR: `DefaultHasher` is not guaranteed to stay the same across Rust versions. Is it important that this is the same for all of the network? Couldn't immediately see that from the libp2p-gossipsub documentation: [`message_id_fn`](https://docs.rs/libp2p-gossipsub/0.46.1/libp2p_gossipsub/struct.ConfigBuilder.html#method.message_id_fn).

jsdanielh replied:
Mmm, I'd say it is important. `libp2p-gossipsub` documentation is not very clear on this. But then, how would you propose to go since we can't use the inner `SipHashers` as they are deprecated?

hrxi replied:
We could use a standard cryptographic hash function, I guess.

jsdanielh replied:
I just changed it to use `sha256`

hrxi replied:
Nice. :)

hrxi commented on network-libp2p/src/discovery/peer_contacts.rs:
```diff
@@ -299,35 +316,43 @@ impl PeerContactBook {
         let info = PeerContactInfo::from(contact);
 
-        if self
+        // A peer is interested to us in two cases:
```
```suggestion
        // A peer is interesting to us in two cases:
```
Looks like grammar but inverts the meaning.

hrxi commented on network-libp2p/src/discovery/peer_contacts.rs:
```diff
@@ -299,35 +316,43 @@ impl PeerContactBook {
+        // A peer is interested to us in two cases:
+        // - I'm configured as a validator, and the peer is also a validator, then that peer is interesting to me
+        //   regardless of the services that are provided by that peer.
+        // - There is a match in the services provided by the peer and our services
```
```suggestion
        // - The services provided by the peer are a superset of the requested services.
```
It was unclear to me what a "match" was supposed to mean. Digging into the source code, it seems to mean "superset".

hrxi commented on network-libp2p/src/discovery/peer_contacts.rs:
```diff
@@ -452,6 +491,69 @@ impl PeerContactBook {
+
+    /// Returns true if an address is valid for dialing.
+    /// It performs basic checks against unsupported addresses.
+    pub fn is_address_valid(&self, address: &Multiaddr) -> bool {
```
```suggestion
    pub fn is_address_dialable(&self, address: &Multiaddr) -> bool {
```
`valid` doesn't really tell me what this function does. Perhaps `dialable` would fit?

hrxi reviewed (commented):
On a more general note, maybe we shouldn't call the wss connections "secure". As far as I know, all of our connections are secure. The difference to wss is that it has two layers that try to secure the connection, the transport and the libp2p protocol itself. I'm not sure what to call it, maybe "secure transport" or simply "secure websocket"? Maybe the latter.

hrxi reviewed (approved):
This looks good to me now. Does it require waiting for a testnet restart? @jsdanielh

jsdanielh commented:
> This looks good to me now. Does it require waiting for a testnet restart? @jsdanielh

Mmm, I'd say that the change in gossipsub might need it

jsdanielh committed:
network-libp2p: Fix flaky test

Fix `ban_peer` test that was sometimes failing due to receiving a
different event (`DhtReady`) than the expected `PeerLeft`.

jsdanielh committed:
network-libp2p: Use `sha256` for the `gossipsub` message ID

Use `sha256` for the `gossipsub` message ID since the default hasher
is not guaranteed to stay the same across Rust versions.
